From 16a1f8fb5f4bae1d12db693035fa0c94f6c8ede7 Mon Sep 17 00:00:00 2001
From: Yannick Fertre <yannick.fertre@foss.st.com>
Date: Thu, 21 Mar 2024 11:04:53 +0100
Subject: [PATCH 1055/1141] drm/panel: simple: fix fill bit per color for DPI
 panel

Check validity of mapping (generate a panic if mapping is null).

Change-Id: I8a8bf04b121ca0482b3a43fb66c780d8337d0fba
Signed-off-by: Yannick Fertre <yannick.fertre@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/368242
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 drivers/gpu/drm/panel/panel-simple.c | 29 ++++++++++++++++++----------
 1 file changed, 19 insertions(+), 10 deletions(-)

diff --git a/drivers/gpu/drm/panel/panel-simple.c b/drivers/gpu/drm/panel/panel-simple.c
index ff7c9e56c77b..92401e1f5dfe 100644
--- a/drivers/gpu/drm/panel/panel-simple.c
+++ b/drivers/gpu/drm/panel/panel-simple.c
@@ -494,18 +494,27 @@ static int panel_dpi_probe(struct device *dev,
 	of_property_read_u32(np, "height-mm", &desc->size.height);
 
 	of_property_read_string(np, "data-mapping", &mapping);
-	if (!strcmp(mapping, "rgb24")) {
+
+	if (mapping) {
+		if (!strcmp(mapping, "rgb24")) {
+			desc->bus_format = MEDIA_BUS_FMT_RGB888_1X24;
+			desc->bpc = 8;
+		} else if (!strcmp(mapping, "rgb565")) {
+			desc->bus_format = MEDIA_BUS_FMT_RGB565_1X16;
+			desc->bpc = 6;
+		} else if (!strcmp(mapping, "bgr666")) {
+			desc->bus_format = MEDIA_BUS_FMT_RGB666_1X18;
+			desc->bpc = 6;
+		} else if (!strcmp(mapping, "lvds666")) {
+			desc->bus_format = MEDIA_BUS_FMT_RGB666_1X24_CPADHI;
+			desc->bpc = 6;
+		}
+	} else {
+		/* No data-mapping node found, set by default bus format & bpc */
+		dev_warn(dev, "%pOF: no data-mapping node found for \"panel-dpi\" binding\n",
+			np);
 		desc->bus_format = MEDIA_BUS_FMT_RGB888_1X24;
 		desc->bpc = 8;
-	} else if (!strcmp(mapping, "rgb565")) {
-		desc->bus_format = MEDIA_BUS_FMT_RGB565_1X16;
-		desc->bpc = 6;
-	} else if (!strcmp(mapping, "bgr666")) {
-		desc->bus_format = MEDIA_BUS_FMT_RGB666_1X18;
-		desc->bpc = 6;
-	} else if (!strcmp(mapping, "lvds666")) {
-		desc->bus_format = MEDIA_BUS_FMT_RGB666_1X24_CPADHI;
-		desc->bpc = 6;
 	}
 
 	/* Extract bus_flags from display_timing */
-- 
2.39.2

